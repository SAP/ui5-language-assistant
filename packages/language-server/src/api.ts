import { resolve } from "path";
import { existsSync } from "fs";

// for use in productive flows
const bundledPath = resolve(__dirname, "..", "..", "dist", "server.js");

const sourcesPath = resolve(__dirname, "server.js");

// We assume that if the `node_modules` directory exists then we are running
// in development mode, We rely on the fact that the bundled .vsix file of the
// VSCode extension excludes the this package's `node_modules` folder (among others).
// Note this logic would not work with yarn 2.0 / npm 8 / pnpm as `node_modules`
// is no longer created by those package managers.
const isDevelopmentRun = existsSync(
  resolve(__dirname, "..", "..", "node_modules")
);

export const SERVER_PATH: string = isDevelopmentRun
  ? /* istanbul ignore else - no tests (yet?) on bundled artifacts */
    sourcesPath
  : bundledPath;
