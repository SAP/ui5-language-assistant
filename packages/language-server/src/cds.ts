import { join } from "path";
import { getLogger } from "./logger";

const metadataCache: { [path: string]: string } = {};

// TODO: multiples services

export async function getCdsMetadata(
  appRoot: string
): Promise<string | undefined> {
  try {
    const cdsRoot = require.resolve("@sap/cds", { paths: [appRoot] });
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const cds = require(cdsRoot);
    const root = cdsRoot.split("/node_modules")[0];
    let csn: any;
    try {
      csn = await cds.load([
        join(root, "app"),
        join(root, "db"),
        join(root, "srv"),
      ]); // TODO: more robust project loading
    } catch (e: any) {
      if (e?.model) {
        csn = e.model;
      } else {
        throw e;
      }
    }
    return cds.compile.to.edmx(csn, {
      version: "v4",
      odataForeignKeys: true,
      odataContainment: false,
    });
  } catch (error: unknown) {
    if (typeof error === "string") {
      getLogger().trace(error);
    } else if (typeof error?.toString === "function") {
      getLogger().trace(error.toString());
    }
  }
  return undefined;
}
