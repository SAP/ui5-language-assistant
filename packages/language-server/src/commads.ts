import { ExecuteCommandParams, createConnection } from "vscode-languageserver";
import {
  QUICK_FIX_STABLE_ID_COMMAND,
  executeQuickFixStableIdCommand,
} from "./quick-fix";

type LSPConnection = ReturnType<typeof createConnection>;

export function executeCommand(
  connection: LSPConnection,
  params: ExecuteCommandParams
): void {
  if (params.arguments === undefined) {
    return;
  }

  switch (params.command) {
    case QUICK_FIX_STABLE_ID_COMMAND: {
      const change = executeQuickFixStableIdCommand({
        // Assumption that this command has the following arguments.
        // We passed them when the command was created.
        textDocument: params.arguments[0],
        quickFixReplaceRange: params.arguments[1],
        quickFixNewText: params.arguments[2],
      });
      connection.workspace.applyEdit({
        documentChanges: change,
      });
      return;
    }
    default:
      return undefined;
  }
}
