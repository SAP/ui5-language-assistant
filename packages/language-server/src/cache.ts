import { readFile } from "fs/promises";
import { fileURLToPath } from "url";
import { sep, normalize, join } from "path";

import { Fetcher } from "api";
import fetch from "node-fetch";
import {
  AppContext,
  ManifestDetails,
  ServiceDetails,
  UI5Framework,
  UI5SemanticModel,
} from "@ui5-language-assistant/semantic-model-types";

import { getMinUI5VersionForXMLFile } from "./manifest-handling";
import { createSemanticModelWithFetcher } from "./ui5-model";

import {
  getCapModelAndServices,
  loadModuleFromProject,
} from "@sap-ux/project-access";

import { parse, merge } from "@sap-ux/edmx-parser";
import { convert } from "@sap-ux/annotation-converter";
import type { Manifest } from "@sap-ux/project-types";
import { getUI5FrameworkForXMLFile } from "./ui5yaml-handling";
import {
  isCapJavaProject,
  isCapNodeJsProject,
} from "@sap-ux/project-access/dist/project/cap";

import {
  findProjectRoot,
  getAppRootFromWebappPath,
} from "@sap-ux/project-access/dist/project/findApps";

import { FileName } from "@sap-ux/project-types";
import { Configuration, UI5Config } from "@sap-ux/ui5-config";
import findUp from "find-up";

const CAP_PROJECT_TYPE = "CAP";
interface CAPProject {
  type: typeof CAP_PROJECT_TYPE;
  kind: CAPProjectKind;
  root: string;
  ui5Config?: UI5Config;
  /**
   * Mapping from service path to service metadata
   */
  services: Map<string, string>;
  apps: Map<string, CachedApp>;
}

const UI5_PROJECT_TYPE = "UI5";

interface UI5Project {
  type: typeof UI5_PROJECT_TYPE;

  root: string;
  ui5Config?: Configuration;
  app?: CachedApp;
}

type CAPProjectKind = "Java" | "NodeJS";
type Project = UI5Project | CAPProject;

async function getUI5Config(
  projectRoot: string
): Promise<Configuration | undefined> {
  try {
    const ui5YamlPath = join(projectRoot, FileName.Ui5Yaml);
    const yamlString = await readFile(ui5YamlPath, "utf-8");
    const ui5Config = await UI5Config.newInstance(yamlString);
    return ui5Config.getConfiguration();
  } catch {
    return undefined;
  }
}

async function getUI5Manifest(
  webappRoot: string
): Promise<Manifest | undefined> {
  try {
    const manifestPath = join(webappRoot, FileName.Manifest);
    const manifestString = await readFile(manifestPath, "utf-8");
    return JSON.parse(manifestString);
  } catch {
    return undefined;
  }
}

async function findAppRoot(path: string): Promise<string | undefined> {
  const manifestJson = await findUp(FileName.Manifest, { cwd: path });
  if (manifestJson) {
    const appRoot = await getAppRootFromWebappPath(manifestJson);
    return appRoot ?? undefined;
  }
  return undefined;
}

const cache: Map<string, Project> = new Map();

interface CachedApp {
  appRoot: string;
  projectRoot: string;
  manifest: Manifest;
  manifestDetails?: ManifestDetails;
  localServices: Map<string, ServiceDetails>;
}

// cache the semantic model creation promise to ensure unique instances per version

const semanticModelCache: Record<
  string,
  Promise<UI5SemanticModel>
> = Object.create(null);

async function getProject(root: string): Promise<Project> {
  const cachedProject = cache.get(root);
  if (cachedProject) {
    return cachedProject;
  }
  const typeAndKind = await getProjectTypeAndKind(root);
  if (typeAndKind.type === CAP_PROJECT_TYPE) {
    const project: CAPProject = {
      ...typeAndKind,
      root,
      services: new Map(),
      apps: new Map(),
    };
    await loadCapServices(project);
    cache.set(root, project);
    return project;
  } else {
    const ui5Config = await getUI5Config(root);
    const project: UI5Project = {
      ...typeAndKind,
      root,
      ui5Config,
    };
    cache.set(root, project);
    return project;
  }
}

async function getProjectTypeAndKind(
  projectRoot: string
): Promise<
  | { type: typeof CAP_PROJECT_TYPE; kind: CAPProjectKind }
  | { type: typeof UI5_PROJECT_TYPE }
> {
  if (await isCapJavaProject(projectRoot)) {
    return {
      type: CAP_PROJECT_TYPE,
      kind: "Java",
    };
  } else if (await isCapNodeJsProject(projectRoot)) {
    return {
      type: CAP_PROJECT_TYPE,
      kind: "NodeJS",
    };
  } else {
    return {
      type: UI5_PROJECT_TYPE,
    };
  }
}

async function findTopProjectRoot(documentPath): Promise<string | undefined> {
  let projectRoot: string | undefined;
  try {
    projectRoot = await findProjectRoot(documentPath, true);
    if (!projectRoot) {
      projectRoot = await findProjectRoot(documentPath, false);
    }
  } catch (e) {
    projectRoot = undefined;
  }
  return projectRoot;
}

export async function getContextForFile(
  uri: string,
  modelCachePath?: string
): Promise<AppContext> {
  const documentPath = fileURLToPath(uri);
  const projectRoot = await findTopProjectRoot(documentPath).catch(
    () => undefined
  );
  const appRoot = await findAppRoot(documentPath);
  const framework = getUI5FrameworkForXMLFile(documentPath);
  const project = projectRoot ? await getProject(projectRoot) : undefined;
  const app = project && appRoot ? await getApp(project, appRoot) : undefined;
  const minUI5Version =
    app?.manifestDetails?.minUI5Version ??
    getMinUI5VersionForXMLFile(documentPath);

  const ui5Model = await getSemanticModel(
    modelCachePath,
    app?.manifestDetails,
    framework,
    minUI5Version
  );

  const services = {};
  for (const [servicePath, service] of app?.localServices ??
    new Map<string, ServiceDetails>()) {
    services[servicePath] = {
      path: service.path,
      convertedMetadata: service.convertedMetadata,
    };
  }

  return {
    services,
    manifest: app?.manifestDetails,
    ui5Model,
  };
}

export async function getSemanticModel(
  modelCachePath: string | undefined,
  manifestDetails: ManifestDetails | undefined,
  framework: UI5Framework,
  version: string | undefined,
  ignoreCache?: boolean
): Promise<UI5SemanticModel> {
  return getSemanticModelWithFetcher(
    fetch,
    modelCachePath,
    framework,
    version,
    ignoreCache
  );
}

// This function is exported for testing purposes (using a mock fetcher)
export async function getSemanticModelWithFetcher(
  fetcher: Fetcher,
  modelCachePath: string | undefined,
  framework: UI5Framework,
  version: string | undefined,
  ignoreCache?: boolean
): Promise<UI5SemanticModel> {
  const frameWorkCacheKey = `${framework || "INVALID"}:${version || "INVALID"}`;
  if (ignoreCache || semanticModelCache[frameWorkCacheKey] === undefined) {
    semanticModelCache[frameWorkCacheKey] = createSemanticModelWithFetcher(
      fetcher,
      modelCachePath,
      framework,
      version
    );
  }
  return semanticModelCache[frameWorkCacheKey];
}
export async function updateManifestData(uri: string): Promise<void> {
  const path = fileURLToPath(uri);
  const projectRoot = await findProjectRoot(path, false).catch(() => undefined);
  const appRoot = await findAppRoot(path);
  if (!projectRoot || !appRoot) {
    return;
  }

  const project = await getProject(projectRoot);
  const app = await loadApp(project, appRoot);
  if (app) {
    if (project.type === CAP_PROJECT_TYPE) {
      project.apps.set(appRoot, app);
    } else {
      project.app = app;
    }
  }
}

export async function updateAppFile(uri: string): Promise<void> {
  const path = fileURLToPath(uri);
  const projectRoot = await findTopProjectRoot(path).catch(() => undefined);
  const appRoot = await findAppRoot(path);
  if (!projectRoot || !appRoot) {
    return;
  }

  const project = await getProject(projectRoot);
  const app = await loadApp(project, appRoot);
  if (app) {
    if (project.type === CAP_PROJECT_TYPE) {
      project.apps.set(appRoot, app);
    } else {
      project.app = app;
    }
  }
}

export async function updatePackageJson(uri: string): Promise<void> {
  const path = fileURLToPath(uri);
  const projectRoot = await findTopProjectRoot(path).catch(() => undefined);
  const appRoot = await findAppRoot(path);
  if (!projectRoot || !appRoot) {
    return;
  }

  cache.delete(projectRoot);

  const project = await getProject(projectRoot);
  const app = await loadApp(project, appRoot);
  if (app) {
    if (project.type === CAP_PROJECT_TYPE) {
      project.apps.set(appRoot, app);
    } else {
      project.app = app;
    }
  }
}

async function loadCapServices(project: CAPProject): Promise<void> {
  try {
    const data = await getCapModelAndServices(project.root);
    const cds = await loadModuleFromProject<any>(project.root, "@sap/cds");
    for (const service of data.services) {
      const metadataContent = cds.compile.to.edmx(data.model, {
        version: "v4",
        service: service.name,
        odataForeignKeys: true,
        odataContainment: false,
      });

      project.services.set(service.urlPath, metadataContent);
    }
  } catch (error) {
    console.log(error);
  }
  return undefined;
}

async function updateCapProject(project: CAPProject): Promise<void> {
  await loadCapServices(project);
  const updatedApps = new Map<string, CachedApp>();
  for (const [, app] of project.apps) {
    const updatedApp = await loadApp(project, app.appRoot);
    if (updatedApp) {
      updatedApps.set(app.appRoot, updatedApp);
    }
  }
  project.apps = updatedApps;
}

export async function updateServiceFiles(uris: string[]): Promise<void> {
  const projectRoots = new Set<string>();
  for (const uri of uris) {
    const path = fileURLToPath(uri);
    const projectRoot = await findTopProjectRoot(path).catch(() => undefined);
    if (projectRoot) {
      projectRoots.add(projectRoot);
    }
  }
  for (const projectRoot of projectRoots) {
    const project = await getProject(projectRoot);
    if (project.type === CAP_PROJECT_TYPE) {
      await updateCapProject(project);
    }
  }
}

async function getApp(
  project: Project,
  appRoot: string
): Promise<CachedApp | undefined> {
  if (project.type === UI5_PROJECT_TYPE) {
    if (project.app) {
      return project.app;
    }
    project.app = await loadApp(project, appRoot);
    return project.app;
  }

  const cachedApp = project.apps.get(appRoot);
  if (cachedApp) {
    return cachedApp;
  }

  const app = await loadApp(project, appRoot);
  if (app) {
    project.apps.set(appRoot, app);
  }

  return app;
}

async function loadApp(
  project: Project,
  appRoot: string
): Promise<CachedApp | undefined> {
  const manifest = await getUI5Manifest(appRoot);
  if (!manifest) {
    return undefined;
  }
  const manifestDetails = await getManifestDetails(manifest);
  const localServices = new Map<string, ServiceDetails>();
  const localServiceFiles = await getLocalServiceData(appRoot, manifest);
  if (localServiceFiles) {
    const metadata =
      project.type === CAP_PROJECT_TYPE
        ? project.services.get(localServiceFiles.path) ??
          project.services.get(localServiceFiles.path.replace(/^\//, ""))
        : undefined;
    if (metadata) {
      // override local service metadata with latest metadata from service
      localServiceFiles.metadataContent = metadata;
    }
    const localService = parseServiceFiles(localServiceFiles);
    if (localService) {
      localServices.set(localService.path, localService);
    }
  }
  const app: CachedApp = {
    projectRoot: project.root,
    appRoot,
    manifest,
    manifestDetails,
    localServices,
  };
  return app;
}

async function getManifestDetails(
  manifest: Manifest
): Promise<ManifestDetails> {
  const flexEnabled = manifest["sap.ui5"]?.flexEnabled;
  const minUI5Version = manifest["sap.ui5"]?.dependencies?.minUI5Version;
  const customViews = {};

  const targets = manifest["sap.ui5"]?.routing?.targets;
  if (targets) {
    for (const name of Object.keys(targets)) {
      const target = targets[name];
      if (target) {
        const settings = (target?.options as any)?.settings;
        if (settings?.entitySet && settings?.viewName) {
          customViews[settings.viewName] = {
            entitySet: settings.entitySet,
          };
        }
      }
    }
  }

  const mainServiceName = getMainService(manifest);
  const mainServicePath = mainServiceName
    ? getServicePath(manifest, mainServiceName)
    : undefined;
  return {
    flexEnabled: flexEnabled,
    minUI5Version: minUI5Version,
    mainServicePath,
    customViews,
  };
}

interface ServiceFiles {
  path: string;
  metadataContent: string | undefined;
  annotationFiles: string[];
}

function parseServiceFiles({
  metadataContent,
  annotationFiles,
  path,
}: ServiceFiles): ServiceDetails | undefined {
  if (!metadataContent) {
    return undefined;
  }

  const myParsedEdmx = parse(metadataContent, "metadata");
  const annotations = annotationFiles.map((file, i) =>
    parse(file, `annotationFile${i}`)
  );
  const mergedModel = merge(myParsedEdmx, ...annotations);

  const convertedMetadata = convert(mergedModel);

  const service: ServiceDetails = {
    path,
    convertedMetadata,
  };

  return service;
}

async function getLocalServiceData(
  appRoot: string,
  manifest: Manifest
): Promise<ServiceFiles | undefined> {
  const mainServiceName = getMainService(manifest);

  if (mainServiceName !== undefined) {
    const metadataContent = await getLocalMetadataForService(
      manifest,
      mainServiceName,
      appRoot
    );
    const annotationFiles = await getLocalAnnotationsForService(
      manifest,
      mainServiceName,
      appRoot
    );
    const path = getServicePath(manifest, mainServiceName) ?? "";

    return {
      path,
      metadataContent,
      annotationFiles,
    };
  }

  return undefined;
}

function getServicePath(
  manifest: Manifest,
  serviceName: string
): string | undefined {
  const dataSources = manifest["sap.app"]?.dataSources;

  if (dataSources && serviceName !== undefined) {
    const defaultModelDataSource = dataSources[serviceName];

    return defaultModelDataSource?.uri;
  }
  return undefined;
}

async function getLocalMetadataForService(
  manifest: Manifest,
  serviceName: string,
  appRoot: string
): Promise<string | undefined> {
  const dataSources = manifest["sap.app"]?.dataSources;

  if (dataSources && serviceName !== undefined) {
    const defaultModelDataSource = dataSources[serviceName];

    const localUri = defaultModelDataSource?.settings?.localUri;
    if (localUri) {
      try {
        const metadataPath = join(appRoot, localUri);
        return await readFile(metadataPath, {
          encoding: "utf8",
        });
      } catch {
        return undefined;
      }
    }
  }
  return undefined;
}

async function getLocalAnnotationsForService(
  manifest: Manifest,
  serviceName: string,
  appRoot: string
): Promise<string[]> {
  const dataSources = manifest["sap.app"]?.dataSources;
  if (dataSources && serviceName !== undefined) {
    const dataSource = dataSources[serviceName];
    const annotationFilePaths = (dataSource?.settings?.annotations ?? [])
      .map((name) => dataSources[name]?.settings?.localUri)
      .filter((path): path is string => !!path);
    if (annotationFilePaths.length) {
      try {
        return await Promise.all(
          annotationFilePaths.map((path) =>
            readFile(join(appRoot, path), {
              encoding: "utf8",
            })
          )
        );
      } catch {
        return [];
      }
    }
  }
  return [];
}

function getMainService(manifest: Manifest): string | undefined {
  const model = manifest["sap.ovp"]?.globalFilterModel ?? "";
  return manifest["sap.ui5"]?.models?.[model].dataSource;
}
