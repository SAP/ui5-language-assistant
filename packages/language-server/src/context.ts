import type {
  AppContext,
  UI5SemanticModel,
} from "@ui5-language-assistant/semantic-model-types";
import { cache } from "./cache";
import { getMinUI5VersionForXMLFile } from "./manifest-handling";
import { getManifestPath } from "./path";
import { getSemanticModel } from "./ui5-model";
import { getUI5FrameworkForXMLFile } from "./ui5yaml-handling";

export const getUI5Model = async (
  documentPath: string,
  modelCachePath?: string
): Promise<UI5SemanticModel> => {
  const minUI5Version = getMinUI5VersionForXMLFile(documentPath);
  const framework = getUI5FrameworkForXMLFile(documentPath);
  const model = await getSemanticModel(
    modelCachePath,
    framework,
    minUI5Version
  );
  return model;
};

export async function getContextForFile(
  documentPath: string,
  modelCachePath?: string
): Promise<AppContext> {
  const ui5Model = await getUI5Model(documentPath, modelCachePath);
  const context: AppContext = {
    ui5Model,
    manifest: cache.getManifestDetails(getManifestPath(documentPath)),
  };
  return context;
}
