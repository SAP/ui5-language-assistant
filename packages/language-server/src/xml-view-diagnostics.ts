import { map, cloneDeep } from "lodash";
import { assertNever } from "assert-never";
import {
  Diagnostic,
  DiagnosticSeverity,
  DiagnosticTag,
} from "vscode-languageserver-types";
import { TextDocument } from "vscode-languageserver-textdocument";
import { DocumentCstNode, parse } from "@xml-tools/parser";
import { buildAst } from "@xml-tools/ast";
import {
  validations,
  DIAGNOSTIC_SOURCE,
} from "@ui5-language-assistant/user-facing-text";
import {
  NonUniqueIDIssue,
  UI5XMLViewIssue,
  validateXMLView,
  XMLViewIssueSeverity,
  defaultValidators,
  validators,
} from "@ui5-language-assistant/xml-views-validation";
import { offsetRangeToLSPRange } from "./range-utils";
import { Context } from "@ui5-language-assistant/context";

export function getXMLViewDiagnostics(opts: {
  document: TextDocument;
  context: Context;
}): Diagnostic[] {
  const documentText = opts.document.getText();
  const { cst, tokenVector } = parse(documentText);
  const xmlDocAst = buildAst(cst as DocumentCstNode, tokenVector);
  const actualValidators = cloneDeep(defaultValidators);
  if (opts.context.manifestDetails.flexEnabled) {
    actualValidators.element.push(validators.validateNonStableId);
  }
  const issues = validateXMLView({
    validators: actualValidators,
    xmlView: xmlDocAst,
    context: opts.context,
  });
  const diagnostics = validationIssuesToLspDiagnostics(issues, opts.document);
  return diagnostics;
}

function validationIssuesToLspDiagnostics(
  issues: UI5XMLViewIssue[],
  document: TextDocument
): Diagnostic[] {
  const diagnostics: Diagnostic[] = map(issues, (currIssue) => {
    const commonDiagnosticPros: Diagnostic = {
      range: offsetRangeToLSPRange(currIssue.offsetRange, document),
      severity: toLspSeverity(currIssue.severity),
      source: DIAGNOSTIC_SOURCE,
      message: currIssue.message,
    };

    const issueKind = currIssue.kind;
    switch (issueKind) {
      case "InvalidBooleanValue":
      case "UnknownEnumValue":
      case "UnknownNamespaceInXmlnsAttributeValue":
      case "UnknownAttributeKey":
      case "UnknownTagName":
      case "InvalidAggregationCardinality":
      case "InvalidAggregationType":
        return {
          ...commonDiagnosticPros,
        };
      case "NonStableIDIssue":
        return {
          ...commonDiagnosticPros,
          code: validations.NON_STABLE_ID.code,
        };
      case "UseOfDeprecatedClass":
      case "UseOfDeprecatedProperty":
      case "UseOfDeprecatedEvent":
      case "UseOfDeprecatedAssociation":
      case "UseOfDeprecatedAggregation":
        return {
          ...commonDiagnosticPros,
          tags: [DiagnosticTag.Deprecated],
        };
      case "NonUniqueIDIssue":
        return {
          ...commonDiagnosticPros,
          relatedInformation: map(
            (currIssue as NonUniqueIDIssue).identicalIDsRanges,
            (_) => ({
              message: validations.NON_UNIQUE_ID_RELATED_INFO.msg,
              location: {
                uri: document.uri,
                range: offsetRangeToLSPRange(_, document),
              },
            })
          ),
        };
      /* istanbul ignore next - defensive programming */
      default:
        assertNever(issueKind);
    }
  });

  return diagnostics;
}

function toLspSeverity(
  issueSeverity: XMLViewIssueSeverity
): DiagnosticSeverity {
  switch (issueSeverity) {
    case "error":
      return DiagnosticSeverity.Error;
    case "warn":
      return DiagnosticSeverity.Warning;
    /* istanbul ignore next - no `Warning` validation yet */
    case "info":
      return DiagnosticSeverity.Information;
    /* istanbul ignore next - no `information` validation yet */
    case "hint":
      return DiagnosticSeverity.Hint;
    /* istanbul ignore next - defensive programming */
    default:
      assertNever(issueSeverity);
  }
}
