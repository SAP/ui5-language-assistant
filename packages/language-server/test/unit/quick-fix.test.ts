import {
  generateModel,
  getFallbackPatchVersions,
} from "@ui5-language-assistant/test-utils";
import { Position, Range } from "vscode-languageserver-types";
import { generate } from "@ui5-language-assistant/semantic-model";
import { UI5SemanticModel } from "@ui5-language-assistant/semantic-model-types";
import {
  DEFAULT_UI5_FRAMEWORK,
  DEFAULT_UI5_VERSION,
} from "@ui5-language-assistant/constant";
import {
  QuickFixStableIdLSPInfo,
  diagnosticToCodeActionFix,
  executeQuickFixFileStableIdCommand,
  executeQuickFixStableIdCommand,
} from "../../src/quick-fix";
import { TextDocument } from "vscode-languageserver-textdocument";
import { getDefaultContext } from "./completion-items-utils";
import { Context } from "@ui5-language-assistant/context";
import { getXMLViewDiagnostics } from "../../src/xml-view-diagnostics";

let appContext: Context;

describe("The @ui5-language-assistant/language-server diagnostics quick fix function", () => {
  let ui5SemanticModel: UI5SemanticModel;
  beforeAll(async function () {
    ui5SemanticModel = await generateModel({
      framework: DEFAULT_UI5_FRAMEWORK,
      version: (
        await getFallbackPatchVersions()
      ).SAPUI5 as typeof DEFAULT_UI5_VERSION,
      modelGenerator: generate,
    });
    appContext = getDefaultContext(ui5SemanticModel);
    appContext.manifestDetails.flexEnabled = true;
  });

  describe("Quick fix code actions", () => {
    it("no quick fixable diagnostics", () => {
      const xmlSnippet = `<mvc:View xmlns:core="sap.ui.core" xmlns:uxap="sap.uxap" xmlns:m="sap.m"
      xmlns:mvc="sap.ui.core.mvc"
      xmlns="sap.ui.commons"
      xmlns:sap.ui.dt="sap.ui.dt">
      </mvc:View>`;

      const xmlTextDoc = TextDocument.create(`dummyUri`, "xml", 0, xmlSnippet);

      const response = diagnosticToCodeActionFix(
        xmlTextDoc,
        [
          {
            message: "",
            range: Range.create(Position.create(0, 0), Position.create(0, 10)),
          },
        ],
        appContext
      );

      expect(response).toBeEmpty();
    });

    it("will get code action", () => {
      const xmlSnippet = `<mvc:View xmlns:core="sap.ui.core" xmlns:uxap="sap.uxap" xmlns:m="sap.m"
      xmlns:mvc="sap.ui.core.mvc"
      xmlns="sap.ui.commons"
      xmlns:sap.ui.dt="sap.ui.dt">
          <m:CheckBox id="dummy-id"></m:CheckBox>
          <m:Button>
              <m:customData>
                  <core:CustomData></core:CustomData>
              </m:customData>
          </m:Button>
          <m:Panel sap.ui.dt:designtime="not-adaptable">
              <m:Dialog>
              </m:Dialog>
          </m:Panel>
      </mvc:View>`;

      const xmlTextDoc = TextDocument.create(`dummyUri`, "xml", 0, xmlSnippet);

      const actualDiagnostics = getXMLViewDiagnostics({
        document: xmlTextDoc,
        context: appContext,
      });
      const response = diagnosticToCodeActionFix(
        xmlTextDoc,
        actualDiagnostics,
        appContext
      );

      expect(response).toMatchSnapshot();
    });
  });

  it("excute Quick Fix Stable Id Command", () => {
    const response = executeQuickFixStableIdCommand({
      documentUri: "dummyUri",
      documentVersion: 1,
      quickFixReplaceRange: Range.create(
        Position.create(0, 0),
        Position.create(0, 10)
      ),
      quickFixNewText: "new text",
    });
    expect(response).toMatchInlineSnapshot(`
      Array [
        Object {
          "edits": Array [
            Object {
              "newText": "new text",
              "range": Object {
                "end": Object {
                  "character": 10,
                  "line": 0,
                },
                "start": Object {
                  "character": 0,
                  "line": 0,
                },
              },
            },
          ],
          "textDocument": Object {
            "uri": "dummyUri",
            "version": 1,
          },
        },
      ]
    `);
  });

  it("execute Quick Fix File Stable Id Command", () => {
    const issues: QuickFixStableIdLSPInfo[] = [
      {
        newText: "newText_1",
        replaceRange: {
          start: Position.create(0, 0),
          end: Position.create(0, 10),
        },
      },
      {
        newText: "newText_2",
        replaceRange: {
          start: Position.create(1, 0),
          end: Position.create(1, 10),
        },
      },
    ];
    const response = executeQuickFixFileStableIdCommand({
      documentUri: "dummyUri",
      documentVersion: 1,
      nonStableIdIssues: issues,
    });
    expect(response).toMatchInlineSnapshot(`
      Array [
        Object {
          "edits": Array [
            Object {
              "newText": "newText_1",
              "range": Object {
                "end": Object {
                  "character": 10,
                  "line": 0,
                },
                "start": Object {
                  "character": 0,
                  "line": 0,
                },
              },
            },
            Object {
              "newText": "newText_2",
              "range": Object {
                "end": Object {
                  "character": 10,
                  "line": 1,
                },
                "start": Object {
                  "character": 0,
                  "line": 1,
                },
              },
            },
          ],
          "textDocument": Object {
            "uri": "dummyUri",
            "version": 1,
          },
        },
      ]
    `);
  });
});
