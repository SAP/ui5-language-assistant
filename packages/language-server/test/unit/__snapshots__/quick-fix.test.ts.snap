// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The @ui5-language-assistant/language-server diagnostics quick fix function Quick fix code actions will get code action 1`] = `
Array [
  Object {
    "command": Object {
      "arguments": Array [
        "dummyUri",
        0,
        Object {
          "end": Object {
            "character": 19,
            "line": 5,
          },
          "start": Object {
            "character": 19,
            "line": 5,
          },
        },
        " id=\\"_IDGenButton1\\"",
      ],
      "command": "ui5_lang.quick_fix_stable_id",
      "title": "Generate an ID",
    },
    "kind": "quickfix",
    "title": "Generate an ID",
  },
  Object {
    "command": Object {
      "arguments": Array [
        FullTextDocument {
          "_content": "<mvc:View xmlns:core=\\"sap.ui.core\\" xmlns:uxap=\\"sap.uxap\\" xmlns:m=\\"sap.m\\"
      xmlns:mvc=\\"sap.ui.core.mvc\\"
      xmlns=\\"sap.ui.commons\\"
      xmlns:sap.ui.dt=\\"sap.ui.dt\\">
          <m:CheckBox id=\\"dummy-id\\"></m:CheckBox>
          <m:Button>
              <m:customData>
                  <core:CustomData></core:CustomData>
              </m:customData>
          </m:Button>
          <m:Panel sap.ui.dt:designtime=\\"not-adaptable\\">
              <m:Dialog>
              </m:Dialog>
          </m:Panel>
      </mvc:View>",
          "_languageId": "xml",
          "_lineOffsets": Array [
            0,
            73,
            107,
            136,
            171,
            221,
            242,
            271,
            325,
            355,
            377,
            434,
            459,
            485,
            506,
          ],
          "_uri": "dummyUri",
          "_version": 0,
        },
        "dummyUri",
        0,
        Array [
          Object {
            "newText": " id=\\"_IDGenButton1\\"",
            "replaceRange": Object {
              "end": Object {
                "character": 19,
                "line": 5,
              },
              "start": Object {
                "character": 19,
                "line": 5,
              },
            },
          },
          Object {
            "newText": " id=\\"_IDGenDialog1\\"",
            "replaceRange": Object {
              "end": Object {
                "character": 23,
                "line": 11,
              },
              "start": Object {
                "character": 23,
                "line": 11,
              },
            },
          },
        ],
      ],
      "command": "ui5_lang.quick_fix_file_stable_id",
      "title": "Generate IDs for the entire file",
    },
    "kind": "quickfix",
    "title": "Generate IDs for the entire file",
  },
  Object {
    "command": Object {
      "arguments": Array [
        "dummyUri",
        0,
        Object {
          "end": Object {
            "character": 23,
            "line": 11,
          },
          "start": Object {
            "character": 23,
            "line": 11,
          },
        },
        " id=\\"_IDGenDialog1\\"",
      ],
      "command": "ui5_lang.quick_fix_stable_id",
      "title": "Generate an ID",
    },
    "kind": "quickfix",
    "title": "Generate an ID",
  },
  Object {
    "command": Object {
      "arguments": Array [
        FullTextDocument {
          "_content": "<mvc:View xmlns:core=\\"sap.ui.core\\" xmlns:uxap=\\"sap.uxap\\" xmlns:m=\\"sap.m\\"
      xmlns:mvc=\\"sap.ui.core.mvc\\"
      xmlns=\\"sap.ui.commons\\"
      xmlns:sap.ui.dt=\\"sap.ui.dt\\">
          <m:CheckBox id=\\"dummy-id\\"></m:CheckBox>
          <m:Button>
              <m:customData>
                  <core:CustomData></core:CustomData>
              </m:customData>
          </m:Button>
          <m:Panel sap.ui.dt:designtime=\\"not-adaptable\\">
              <m:Dialog>
              </m:Dialog>
          </m:Panel>
      </mvc:View>",
          "_languageId": "xml",
          "_lineOffsets": Array [
            0,
            73,
            107,
            136,
            171,
            221,
            242,
            271,
            325,
            355,
            377,
            434,
            459,
            485,
            506,
          ],
          "_uri": "dummyUri",
          "_version": 0,
        },
        "dummyUri",
        0,
        Array [
          Object {
            "newText": " id=\\"_IDGenButton1\\"",
            "replaceRange": Object {
              "end": Object {
                "character": 19,
                "line": 5,
              },
              "start": Object {
                "character": 19,
                "line": 5,
              },
            },
          },
          Object {
            "newText": " id=\\"_IDGenDialog1\\"",
            "replaceRange": Object {
              "end": Object {
                "character": 23,
                "line": 11,
              },
              "start": Object {
                "character": 23,
                "line": 11,
              },
            },
          },
        ],
      ],
      "command": "ui5_lang.quick_fix_file_stable_id",
      "title": "Generate IDs for the entire file",
    },
    "kind": "quickfix",
    "title": "Generate IDs for the entire file",
  },
]
`;
